packages/styled-components/src/types.ts

```diff
@@ -285,3 +285,12 @@ export type CSSProp = Interpolation<any>;
 export type NoInfer<T> = [T][T extends any ? 0 : never];
 
 export type Substitute<A extends object, B extends object> = FastOmit<A, keyof B> & B;
+
+type AnyIfEmpty<T extends object> = keyof T extends never ? any : T;
+
+export interface ThemeProps<T> {
+  theme: T;
+}
+
+export type ThemedStyledProps<P, T> = P & ThemeProps<T>;
+export type StyledProps<P> = ThemedStyledProps<P, AnyIfEmpty<DefaultTheme>>;
```

Comment:
When I upgraded from v5 to v6, I was able to replace `ThemedStyledProps<P, T>` with `P & ExecutionContext` in my codebase. 

If we do want to add some aliases for the old v5 types to make upgrading easier, we could map them to the new types like this:

```ts
export type ThemedStyledProps<P> = P & ExecutionContext;
```

Reply:
Sure, that would be fine, but it would lead to inconsistencies in the type arguments of `StyledProps`.


I think it's better to wait for this to be addressed.
https://github.com/styled-components/styled-components/pull/4103/files#diff-adc775db811d749ad367a5fd61454d85f453f549d088fd97ea1575c3eb3290f4R75-R76

(and, until then, the old-style types should be defined anyway.)